### ISSUE TITLE
C#: infinite recursive setter should be warned at compile time

### ISSUE THREAD

**Issue by srivatsn** [dotnet/roslyn-analyzers#777]
Ported from https://github.com/dotnet/roslyn/issues/7695 - reported by @zhangtairan

---

For example:

```
    class Test
    {
         public int Abc
         {
             set
             {
                 this.Abc = value;
             }
         }
    }
    Test test = new Test();
    test.Abc = 1;
```

This behavior will cause StackOverflowException, why not disable infinite recursive setter by the compiler?
Thanks!

**Comment by mavasani** [comment]
@jasonmalinowski Is this similar to the issue we discussed at last design meeting? Seems it has been requested few times now...

**Comment by jasonmalinowski** [comment]
@mavasani Yep, same suggestion!

**Comment by Evangelink** [comment]
@mavasani Could we extend the rule `CA2245` - `AvoidPropertySelfAssignment` to include this scenario? Or do you think this should be a separate rule?

**Comment by Evangelink** [comment]
I am also wondering whether you think it's best to:

1. Register a syntax action (I will need to access the symbol to compare the declared property and the assigned property). I really don't think that's the good option.

2. Register an operation action on the `MethodBody` and loop through assignment children.

3. Register an operation action on the `SimpleAssignment` and look for cases where some ancestor is the setter property

**Comment by mavasani** [comment]
Related issue https://github.com/dotnet/roslyn-analyzers/issues/1478 - we should close one of them as dupes.

> Could we extend the rule CA2245 - AvoidPropertySelfAssignment to include this scenario? Or do you think this should be a separate rule?

I think probably good to be a separate rule as CA2245 flags a potential functionality bug, while this one is a reliability bug.

**Comment by Evangelink** [comment]
This case seems a lot easier to fix that the "general" infinite recursion. Maybe we could add a first version just checking for the simple case and then replace it with the more general one. This way some value is already quickly provided :)

Note that I will try to have a look at the dataflow rule to see if I manage to do something.

**Comment by mavasani** [comment]
I like approach (3). I am not sure you need any operation tree walking. You can potentially do the following:

1. Register an OperationBlockStartAction, which checks if the block's [OwningSymbol](https://github.com/dotnet/roslyn/blob/0a2a70387ecc150cf222341f87a60b108931f4ae/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L1008-L1011) is a property setter with a [MethodKind](https://github.com/dotnet/roslyn/blob/0a2a70387ecc150cf222341f87a60b108931f4ae/src/Compilers/Core/Portable/Symbols/MethodKind.cs#L75-L78) check. Please confirm if the owning symbol that is provided for this context is indeed the property setter and not the property symbol itself, my intuition says it should be the former.
2. If the check above succeeds, then register a nested OperationAction for `SimpleAssignment` that validates if the assignment target is an `IPropertyReferenceOperation` whose target property is the same as setter's `AssociatedSymbol` and it's [Instance](https://github.com/dotnet/roslyn/blob/0a2a70387ecc150cf222341f87a60b108931f4ae/src/Compilers/Core/Portable/Operations/OperationInterfaces.xml#L751-L755) is `IInstanceReferenceOperation` with [InstanceReferenceKind.ContainingTypeInstance](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs#L10-L14)

**Comment by mavasani** [comment]
> This case seems a lot easier to fix that the "general" infinite recursion. Maybe we could add a first version just checking for the simple case and then replace it with the more general one. This way some value is already quickly provided :)

Yes, absolutely. I would also suggest starting with just implementing just the property setter case first, and then enhancing it for more complex method call infinite recursion scenarios.
